# function-built-in-class-simple-float
> [TABLE OF CONTENTS](https://nbviewer.jupyter.org/github/SeanOhAileasa/fubar-python/blob/main/fubar-python.ipynb#top)
***

```python
class float([x])
```
> [Python Documentation](https://web.archive.org/web/20200503112950/https://docs.python.org/3/library/functions.html#float) [0]

- store fractional numbers [1.25]

In [1]:
0.000005

5e-06

- defined either in standard decimal notation or in exponential notation [1.25] - exponential notation - e or E [1.26]

$5e-6 = 5x10^{-6}$

In [2]:
1.4*(10**6)

1400000.0

$1.4e6 = 1.4x10^{6}$

```python
isinstance(object, classinfo)
```
> [Python Documentation](https://web.archive.org/web/20200502105905/https://docs.python.org/3/library/functions.html#isinstance)

In [3]:
isinstance(0.000005,object)

True

In [4]:
isinstance(5e-6,object)

True

In [5]:
isinstance(0.000005,float)

True

In [6]:
isinstance(5e-6,float)

True

```python
class type(object)
```
> [Python Documentation](https://web.archive.org/web/20200503111504/https://docs.python.org/3/library/functions.html#type)

In [7]:
type(0.000005)

float

In [8]:
type(5e-6)

float

- floating-point arithmetic - precision is limited - can cause equality tests to be unstable [1.26]

In [9]:
1.0+0.2==3.0

False

- behaviour not unique to python - due to the fixed-precision format of the binary floating-point storage used by most scientfic computing platforms - all programming languages using floating-point numbers store them in a fixed number of bits which leads some numbers to be represented only approximately [1.26]

```python
class str(object='')
class str(object=b'', encoding='utf-8', errors='strict')
```
> [Python Documentation](https://web.archive.org/web/20200503112950/https://docs.python.org/3/library/functions.html#func-str)

```python
str.format(*args, **kwargs)
```
> [Python Documentation](https://web.archive.org/web/20200502135931/https://docs.python.org/3/library/stdtypes.html#str.format)

In [10]:
"0.1={0:.17f}".format(0.1)

'0.1=0.10000000000000001'

In [11]:
"0.2={:.17f}".format(0.2)

'0.2=0.20000000000000001'

In [12]:
"0.3={:.17f}".format(0.3)

'0.3=0.29999999999999999'

- accustomed to thinking of numbers in decimal - base-10 notation - each fraction must be expressed as a sum of powers of ten [1.26]

In [13]:
1/8

0.125

$1/8 = 0.125 = (1x10^{-1}) + (2x10^{-2}) + (5x10^{-3})$
- computers usually store values in binary notation - each number is expressed as a sum of powers of two [1.26] - base-2 representation is written with a subscript 2 indicating binary notation

$0.001_{2} = (0x2^{-1}) + (0x2^{-2}) + (1x2^{-3})$
- happens to be one number which both binary and decimal notation can represent in a finite number of digits

$0.125 = 0.001_{2}$
- in the familiar base-10 representation of numbers - numbers that cannot be expressed in a finite number of digits - dividing one by three - the threes go on forever - to truly represent this quotient the number of required digits is infinite [1.27]

In [14]:
1/3

0.3333333333333333

- similar to the base-10 representation of numbers there are numbers for which binary representations require an infinite number of digits

$1/10 = 0.00011001100110011..._{2}$
- just as decimal notation requires an infinite number of digits to perfectly represent 1/3 - binary notation requires an infinite number of digits to represent 1/10 - python internally truncates these representations as 52 bits beyond the first non-zero bit on most systems - this rounding error of floating-point values is a necessary evil of working with floating-point numbers - best way to deal with it is to always keep in mind that floating-point arithmetic is approximate and never rely on exact equality tests with floating-point values [1.27]

***
# END